### ISSUE TITLE
When loading a saved model, if keras_nlp is not imported, an untrained model is created

### ISSUE THREAD

**Issue by elbamos** [keras-team/keras-hub#717]
**Describe the bug**
I built a model using the TokenAndPositionEmbedding and TransformerEncoder layers, then saved the model. If I reload the model after importing `keras_nlp`, everything is fine. If, however, I reload the model without importing `keras_nlp`, I get a model that produces much worse performance on the same data. The restored model performs like a model that has never been trained. It is as if - and this is what I think is happening - the weights for something inside the TransformerEncoder layers are not being restored, and instead new weights are being created. 

The reason this is particularly problematic is that I would like to deploy my trained model on Vertex, where I don't think `keras_nlp` is available.

**To Reproduce**
[ReproductSaveIssue.ipynb.zip](https://github.com/keras-team/keras-nlp/files/10573928/ReproductSaveIssue.ipynb.zip)

**Expected behavior**
The model should be restored properly. 

**Additional context**
This issue may be related: https://github.com/keras-team/keras-nlp/issues/219

**Would you like to help us fix it?**

How can I help?

**Comment by elbamos** [comment]
@jbischof wanted to make sure you saw this, its a minimally reproducible example of the saving issue I tried to raise a few days ago.

**Comment by mattdangerw** [comment]
Here's a colab version of the repro -> https://colab.research.google.com/gist/mattdangerw/9e14379d76991bf61687348cbebb957b/reproductsaveissue.ipynb

This is a good catch! We don't have testing for saving a model in an environment where `keras_nlp` is available, and then loading in an environment where it is not.

I am not totally sure what the correct fix here is, let us get back to you! At the highest level, we do want our models easily usable with Vertex, so we should make sure we have a good path of support there.

**Comment by elbamos** [comment]
Am I wrong that if a compiled keras model is saved with traces, it _should_ be runnable on another system without the python code for the modules? The model does correctly convert to tflite in portable form.

**Comment by mattdangerw** [comment]
> Am I wrong that if a compiled keras model is saved with traces, it should be runnable on another system without the python code for the modules? The model does correctly convert to tflite in portable form.

Yes that is correct, though the best way to get a model that is purely "traced" would probably be with the low level [tf.saved_model.save](https://www.tensorflow.org/api_docs/python/tf/saved_model/save) API. This will get you a fully traced graph without any of the python code. `model.save()` is actually a bit of a hybrid, it will attempt to revive python objects and some model traces on top, so for `model.save()`, having the KerasNLP library imported in the deployment environment is probably a good idea.

Random note, this confusion is a big pain point, and there is active work at the core Keras level to clean up this picture. There was some info about future directions in our latest community meeting -> https://drive.google.com/file/d/1wWg_5Eu0ODQhXdLPMPpqt2EiiBg2um1O/view?usp=sharing

So hopefully soon a less confusing picture around all of this!

**Comment by jayam30** [comment]
Hey!
I want to solve this bug

**Comment by maitry63** [comment]
Hi @elbamos , I tried reproducing this with KerasHub using a model containing `TokenAndPositionEmbedding` and `TransformerEncoder`.

Training and saving :

```
import numpy as np
import keras
import keras_hub

inputs = keras.Input(shape=(16,), dtype="int32")

x = keras_hub.layers.TokenAndPositionEmbedding(
    vocabulary_size=1000,
    sequence_length=16,
    embedding_dim=32,
)(inputs)

x = keras_hub.layers.TransformerEncoder(
    intermediate_dim=64,
    num_heads=4,
)(x)

x = keras.layers.GlobalAveragePooling1D()(x)
outputs = keras.layers.Dense(1, activation="sigmoid")(x)

model = keras.Model(inputs, outputs)

model.compile(
    optimizer="adam",
    loss="binary_crossentropy",
)

x_train = np.random.randint(0, 1000, size=(1000, 16))
y_train = np.random.randint(0, 2, size=(1000, 1))

model.fit(x_train, y_train, epochs=3)

sample = x_train[:10]
preds_before = model.predict(sample)

np.save("preds_before.npy", preds_before)
np.save("sample.npy", sample)

model.save("test_model.keras")

print("Saved model.")
```

Loading in a fresh process without using `keras_hub`:

```
import numpy as np
import keras

sample = np.load("sample.npy")
preds_before = np.load("preds_before.npy")

model = keras.models.load_model("test_model.keras")

preds_after = model.predict(sample)

print("max abs diff:",
      np.max(np.abs(preds_before - preds_after)))

print("mean abs diff:",
      np.mean(np.abs(preds_before - preds_after)))
```

Output:

```
max abs diff: 0.0
mean abs diff: 0.0
```

So, the model loads successfully and produces identical predictions after save, means the weights and model state were restored correctly.

Since `KerasNLP` has migrated into `KerasHub`, and this issue is no longer reproducible with the latest versions, I think it's reasonable to close this issue. If anyone can still reproduce the problem on current versions, please feel free to open a new issue with an updated reproducible example.

Thank you !
